### ISSUE TITLE
Discussion: Namespaces

### ISSUE THREAD

**Issue by mansellan** [WaynePhillipsEA/twinbasic#111]
**Is your feature request related to a problem?**
In VB6 / VBA, all referenced types are imported into a single global namespace. Collisions can be resolved be prefixing the library name (`Excel.Range` vs `Word.Range`), and methods in standard modules can be disambiguated using the module name. But that's as much as you get. This leads to difficulty in using common words for types, and huge intellisense lists.

**Describe the solution you'd like**
Ideally, namespaces similarly to VB.Net, but it's... complicated. The current behaviour is deeply ingrained, part of the DNA of the language. All the standard library functions, and any referenced library, are not going to be namespace-aware not matter what happens with twinBASIC. At the same time, if the hope is to evolve the language such that greenfield projects can be written more expressively than was possible in the past, true namespace support would be a huge win. Naming things is hard, doubly so if you only have one namespace.

This issue is very much intended as a discussion, I'm really not sure whether a solution exists. I don't believe this is an urgent problem, VB6 managed without namespaces. It is something to consider for the future.

Some things that a candidate solution should consider:
1. It should not require additional cognitive load when using or writing in the VB6 style (reference a lib, use its types).
2. It must be compatible with COM and DLL exports (see #22)
3. What impact would any proposed solution have on type resolution? What happens if a namespace level collides, e.g. `MyApp.VBA`?
4. What level should it act upon? Component/File/Project?

**Describe alternatives you've considered**
Some have taken to [simulating namespaces]9https://bytecomb.com/organizing-code-with-namespaces-in-vba/) using predeclared classes and/or type hierarchies, but this is an imperfect compromise and adds bloat to a codebase. 

**Additional context**
An initial proposal, I'm sure there are problems with it:

1. Add an `Option Namespace` - without this a component uses existing behaviour (opt-in).
If set:
2. Library functions are NOT automatically imported (either from `VBA` or any referenced libs)
3. Namespaces can be defined: `Namespace MyApp.Models.Shipping`
4. Namespaces can be imported: `Import VBA`, `Import MSForms`, `MyApp.Models.Shipping`
5. Namespaces can be aliased: `Import MSForms As Forms`

Not sure what this would mean for exported types, although in other issues the use of attributes is being considered (see #22). I _think_ this might help with the issue of existing libraries though - I could envisage writing a "wrapper library" for e.g. MSForms that exposed its types in a series of smaller namespaces.

**Comment by WaynePhillipsEA** [comment]
I think this is a good long-term goal.  FWIW, the twinBASIC compiler/resolver is actually already namespace-aware, but just has the one implicit 'global' unnamed namespace defined.

**Comment by WindowStations** [comment]
https://www.youtube.com/watch?v=N9QAgIpwS5A
The video I posted above shows vb.net namespaces consumed by VB6 as nested class buckets (a term used by MS before NET). See the last minute of the video for namespaces.

**Comment by Greedquest** [comment]
> An initial proposal, I'm sure there are problems with it:
> 
> 1. Add an `Option Namespace` - without this a component uses existing behaviour (opt-in).
>    If set:
> 2. Library functions are NOT automatically imported (either from `VBA` or any referenced libs)
> 3. Namespaces can be defined: `Namespace MyApp.Models.Shipping`
> 4. Namespaces can be imported: `Import VBA`, `Import MSForms`, `MyApp.Models.Shipping`
> 5. Namespaces can be aliased: `Import MSForms As Forms`
> 
> Not sure what this would mean for exported types, although in other issues the use of attributes is being considered (see #22). I _think_ this might help with the issue of existing libraries though - I could envisage writing a "wrapper library" for e.g. MSForms that exposed its types in a series of smaller namespaces.

I've been having a think about this and in light of the new multi-file support I think I'd like to suggest a slightly different approach.

If a namespace is just a list of names that refer to things that can be accessed (without qualification) from a certain scope, then VB* already has 3 hierarchical levels of namespace<sup>1</sup>:

<sub>_(most enclosing)_</sub>

1. *Project namespace* - true global, defines the set of names accessible from all contexts in a project, think Public Subs in a module, class names or names imported from referenced libraries.
2. *Component namespace* - within a module or class there are methods and data that may only be accessible to code running inside that component - e.g. private member variables of a class.
3. *Local namespace* - names accessible to code inside a function like locally `Dim`med variables

<sub>_(most enclosed)_</sub>

Let me explain a little more:

### Scopes and Namespaces in VB*
When I say that namespaces "enclose" one another/can be "enclosed", I mean, for example, the _Component namespace_ - list of names defined for code in that component - is a superset of the enclosing _Project namespace_, so all code defined in a component can use names listed in both the *Component* and *Project namespaces*. Or in other words, the component namespace of `MyClass` effectively contains all the methods, variables, types of `MyClass` as well as `VBA.CLngPtr` or `Collection` or global variables etc., but *not* local variables inside one of MyClass' methods.

It's also worth mentioning that methods of a Class with the `Public` access modifier are not in the Project namespace even if that class is; e.g. I can `Dim x As MyClass` in any location in my project, so `MyClass` is in the _Project namespace_, however the `Foo` method in a call  like`x.Foo` is qualified by "x.", so `Foo` is not in the project namespace (even if it is declared with a project-wide _scope_ so can be called from anywhere). Contrast this to if `Foo` was declared with the Public access modifier in a standard module `MyModule`. Now both `MyModule` and `Foo` are in the _Project namespace_, since `MyModule.Foo` and `Foo` on its own are both valid. In this way the Public access modifier has 2 jobs:
1. In a class *or* standard module, it declares a name with project-wide scope
2. Only in a standard module, `Public Sub Foo` _also_ exports the name `Foo` to the enclosing project namespace whereas for a class it does not.

In general in VB*, 
 -  you can export from *Component namespace* -> *Project namespace* with the Public access modifier in a standard method. 
 - There is no way to export from *Local namespace* to *Component namespace* or declare a name in a local scope and have it accessible from an enclosing scope (no `nonlocal a As Long`). 

Add reference will dump the exported names from the referenced lib into project namespace with shadowed member resolution.

<sub><sup>1</sup>There is a "0th" namespace, maybe call it the _Exposed namespace_ which in VBA is populated with PublicNotCreatable classes, predeclared classes and public module members without Option Private Module, basically the stuff visible to the outside world. I'm gonna ignore it for now</sub>

---

Ok, definitions aside, **what I would propose for twinBASIC** is a 4th namespace in between the Project and Component namespaces; a new _File namespace_.

What I mean by this is that rather than declaring Modules and Class names in the project namespace by default, and rather than exporting `Public Subs` defined in a module to the project namespace, all the information in a .twin file goes into an intermediate File namespace.

So before in VB* the following code:

```vba
'''scratchpad.bas
Private Const MODULE_NAME As String = "scratchpad.bas" 'component scope, visible to all members of this module

Private Function CallMe() 'this method is component scope too, as other methods in the "scratchpad.bas" module can see it
	Static callCount As Long 'Declared with Local Scope, not visible outside this function
    Debug.Print MODULE_NAME; "was called - tot calls ="; callCount
	callCount += 1
End Function

Public Sub RunCode() 'the Public keyword *exports* something from component to file/project namespace
	CallMe 'CallMe is private sure, but this local scope can access enclosing component namespace as well as local namespace
End Sub

'''MyClass.cls
Public Sub Foo()
	...
End Sub
```
... would define the following additions to/new namespaces:
```vba
ProjectNamespace(ExampleProject) += [scratchpad, RunCode, MyClass] 'and by extension scratchpad.RunCode & MyClass.Foo
	ComponentNamespace(scratchpad.bas) = [MODULE_NAME, CallMe, RunCode]
		LocalNamespace(Sub CallMe) = [callCount]
		LocalNamespace(Sub RunCode) = []

	ComponentNamespace(MyClass.cls) = [Foo]
		LocalNamespace(Sub Foo) = [...]
	
	'...
```
_note we're adding to `ProjectNamespace`, polluting it_ 

---
tB meanwhile would have

```vba
'file1.twin
Module scratchpad
	  Private Const MODULE_NAME As String = "scratchpad.bas"
	  
	  Private Function CallMe()
		  Static callCount As Long
	      Debug.Print MODULE_NAME; "was called - tot calls ="; callCount
		  callCount += 1
	  End Function

	  Public Sub RunCode()
		  CallMe
	  End Sub
End Module

'''file2.twin
'file scope - python can have code here but VB* can't, modules suffice
Module scratchpad 'no name clash as this is file2's file namespace
    'component scope - in VB* we can't write code that executes here I don't think
	Sub test()
        'local scope - VB* always runs code in local scope, but it can access members of enclosing scopes
        Debug.Print CLngPtr(12) 'fine, VBA lib added in project references, meaning it is in project namespace
                                'and local scope can access project namespace
		RunCode 'Error, name not defined, parallel file1.twin file namespace inaccessible from here
    End Sub
End Module
```

... which defines:

```vba
ProjectNamespace(ExampleProject) += [] 'still populated with VBA.Collection, Scripting, Excel etc, but no user code... (yet)
	FileNamespace(file1.twin) = [scratchpad, RunCode] 'this would have gone into true global ProjectNamespace 
		ComponentNamespace(Module scratchpad) = [MODULE_NAME, CallMe, RunCode] 'below as before
			LocalNamespace(Sub CallMe) = [callCount]
			LocalNamespace(Sub RunCode) = []

	FileNamespace(file2.twin) = [scratchpad, Test] 'each file has its own FileNamespace so name collisions in user code are not an issue
		ComponentNamespace(Module scratchpad) = [Foo]
			LocalNamespace(Sub Test) = [] 'but again, although LocalNamespace = [], code running in local scope has access
				                          ' to the sum of all enclosing namespaces, i.e.:
				                          ' AccessibleNamespace = ProjectNamespace + FileNamespace + ComponentNamespace + LocalNamespace
```

Each file has its own namespace which is not visible to the other file, meaning `Public Sub RunCode` in file1 is not project-wide and so not accessible from file2 - similarly classes would not be instantiable from other `.twin` files in a project. To get around this there would be 2 options:

 1. `.bas` and `.cls` files would be supported which allow for the existing behaviour by being added directly to the *Project namespace*. Combined with the fact that project references are in the project namespace, this would be identical to existing behaviour.
 2. For `.twin` files meanwhile, a new `import` statement can explicitly bring names from another file's _File namespace_ into any enclosing namespace<sup>2</sup>

```vba
'''file1.twin
from file2 import Foo 'into file namespace

Module m
   from file2 import Bar 'into component namespace
   Sub t()
       import file2.Baz 'into local namespace
       'sees Baz, Bar & Foo`
   End Sub
   Function f()
       'can see Bar & Foo but not Baz
   End Function
End Module

Class c
	Sub x()
		'can see Foo in file namespace but not Bar or Baz
	End Sub
End Class
```

This behaviour is very similar to python (except python lets you run code in any scope and can have functions within functions so is a bit more complicated).

To start with `import file2` can put everything from file2's file namespace (or exposed namespace?) into file1's file namespace

<sub><sup>2</sup> Well not the full file namespace, I think `Public/Private` on a component could dictate whether it is importable by other files in the project. This would hopefully extend to other projects and allow import from other tB code, compiled dlls are another issue</sub>

---

### Summary

It think as @mansellan says: 

 > The current behaviour is deeply ingrained, part of the DNA of the language

and because of that I respectfully disagree with the suggested proposal:

 > 2. Library functions are NOT automatically imported (either from `VBA` or any referenced libs)
 
I think the project references should work as they always have by bringing the exposed members into the project namespace where they are accessible unqualified from all scopes using shadowed member resolution, I think this is unavoidable for backwards compatibility "the DNA". On top of this, users should be able to add `[Legacy]StandardModule` and `[Legacy]ClassModule` files to their project (they could appear as .bas & .cls in the virtual filesystem although based on how you handle file metadata like attributes maybe you want a different file extension - see #162). These Legacy components also populate the project namespace like in VB*.

Consequently I also disagree with:

> 1. Add an `Option Namespace` - without this a component uses existing behaviour (opt-in).

... as this will be accounted for by supporting legacy files for the standard behaviour, and:

> 3. Namespaces can be defined: `Namespace MyApp.Models.Shipping`

Namespace should not be a keyword - I think the file-level boundary is sufficient to delineate where these new namespaces fit it. Namespaces will be automatically named based on filename and directory structure, e.g.

    /Sources/Models/Shipping.twin 'in MyApp

could define the namespace `MyApp.Models.Shipping`, along with relative imports like if 2 twin files are adjacent in the directory structure, they don't then need the `MyApp.Models`.

As for the solution criteria:


> 1. It should not require additional cognitive load when using or writing in the VB6 style (reference a lib, use its types).

It could in fact be identical

> 3. What impact would any proposed solution have on type resolution? What happens if a namespace level collides, e.g. MyApp.VBA?

Existing namespace resolution goes Local > Component > Project > References in order - with the exception of language keywords - so I think just slip File before Project and all is well.

> 4. What level should it act upon? Component/File/Project?

Namespaces already exist, but the new one would act at a file level

> 2. It must be compatible with COM and DLL exports (see #22)

Already the *Project namespace* was not the same thing as the _Exposed namespace_ which does not include all the stuff imported into the project namespace by adding references, or things like Option Private Module etc.

Would require a separate issue to discuss probably. One drawback with this approach is that it does not allow you to use namespaces directly with referenced dlls since the add reference behaviour is project-wide. You could write a thin tB wrapper to namespace-ify a reference by creating a number of `.twin` files which each exposes some subset of the reference's functionality. 

However compiling tB code to COM compatible would nuke the structure (though not break anything, since the compiler *is* aware of what `x` might be referring to), unless there's a way in COM to support namespaces - maybe it's just the technology. The workaround is not compiling to COM dll -  if you could reference other tB source code like an uncompiled/half-compiled twinproj. In one possible implementation you could basically simulate having one project's source in an adjacent directory in the virtual filesystem. VBA doesn't need to name its namespaces because they are all implicitly exported to one another rather than implicitly imported by one another, so encapsulation is achieved through access modifiers.


# TL;DR
 - VB* already have a concept of hierarchical implicit namespaces;
    0. (Exposed namespace) 
	1. Project namespace
	2. Component namespace
	3. Local namespace
 - I propose a new File namespace that sits between Project and Component
 - This will shield the Global Project Namespaces from being polluted by components defined in that project's .twin files
 - To access another .twin file's namespace an implicit import of that namespace or names within it is required
 - Alternatively .bas & .cls files can be used to add components to the project namespace
 - Other libraries written in tB can be imported more flexibly, perhaps by placing their source in an adjacent folder in the virtual filesystem.
 - Compiled COM dlls can be added as references and continue to populate the project namespace, simulating VB* behaviour exactly.


---
_Hope that's clear, let me know if I can clarify bits, much inspired by [Python's approach](https://realpython.com/python-namespaces-scope)_

**Comment by mansellan** [comment]
@Greedquest most insightful. I need to re-read it a few times to make sure I follow everything, and should probably read up on what Python does too. I'm very heavily .Net biased at the moment, and especially C#.

From what I understand though, it does seem like you might be conflating namespaces with scopes. The hierarchy you describe is what I understand to be scopes - the symbols that are available at a particular level. That's not quite the same as a namespace. IIUC, VBx doesn't really understand namespaces, but it does have scoping rules. If I reference a library, it imports to the global (only) namespace. If there are naming clashes, I have limited ways of fixing it to get the code to compile - I can prefix the library name, or with user code I can prefix the module name if it's a standard module.

Adding an additional scoping level would certainly help to avoid clashes, but it seems like a partial solution. In .Net, namespaces can be chosen explicitly [1], allowing ultimate freedom. For example, at work I have some libraries which import into the `System.Linq` namespace. This is a deliberate (some would say reckless!) choice, which means that as soon as someone references the library, any time they use the (built-in) `System.Linq` namespace, they also get all of the overloads I added to it. That level of flexibility doesn't seem possible with your proposal?

[1] In practice, they're often not quite so carefully chosen. The default behaviour for Visual Studio when adding a new class is to put it into a namespace based on it's location. Many developers don't bother to think about changing it.

**Comment by WindowStations** [comment]
Mansellan, I ripped a bunch of net class clones with reflection and put them into namespaces on the NET side, making them comvisible, but I don't seem to have any real bloat like NET.  Although it's fair to say that all the enums/classes can be seen and instanced, I don't think that's the real issue with vba/vb6 namespace awareness.  I like having the extra features available in intellisense, but the class buckets don't actually expose anything outside of the hierarchy.  The problem that I see is that NET classes can be used as an instance, but sometimes used directly (imported) like this:  Process.Start.   Instead of the sample, [https://bytecomb.com/organizing-code-with-namespaces-in-vba/](url) , use a creatable class in VB6 and simply change the module declaration to an empty property returning the class or New Class as needed.

Change this:
Public IO As New nsIO

To this:
Public Property Get IO() As nsIO
End Property

That exposes the same NET functionality to VB6 depending on the constructors used on the NET side.

**Comment by bclothier** [comment]
FWIW, any discussion around scoping, of which the namespace is a part of, must be in compliance with the MS-OAUT and MS-VBAL specifications at least for publicly exposed entities. 

For references:

[MS-OAUT: 2.2.49.2 IDL Automation Scope](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-oaut/ae4d27fa-faaa-4d5f-b0f9-15bfbaaae9b4)
[MS-OAUT: 2.2.49.10 Referencing External Types](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-oaut/ca898dc2-450e-4e2e-983e-1eb1f0a68a45)
[MS-VBAL: 4 VBA Program Organization](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/1a224433-0b53-4ebd-a422-1232a25ed701)
[MS-VBAL: 5.6.10 Simple Name Expressions](https://docs.microsoft.com/en-us/openspecs/microsoft_general_purpose_programming_languages/ms-vbal/e3af3398-f090-40db-ade6-de3a93589c76)

For me personally, I look at namespaces as a logical containers. It need not to correspond to the filesystem structure, though it is common to do so. I'm a bit ambivalent about using C-style includes, of which the Python sample above strikes me as being similar. One reason for that is because this hinders the ability to refactor the code. If the namespace is not a code element, then it's hard to refactor without breaking the code. Consider that refactoring could be:

1) moving a member from one type or module to another type/module
2) moving a type's place in the namespace (e.g. add a new subfolder)
3) moving a type into a new library altogether

And a good refactor would take in account of all the call sites & usages of the moved type so that the code is not left in a broken state. If not making namespace a code element makes it hard to achieve, then I would be disinclined towards any solutions that doesn't make namespace a code element. 

Though the goal is full compatibility, we need to recognize that how the type library is designed, there's a perverse incentive for making huge type libraries huge & bloated. One huge factor is the reluctance to add several references, and in VBA projects, the reference's full path are used which also causes problem for caching & porting to other computers. If the namespaces are to be visible to non-tB consumers, then it needs to somehow solve that problem. One possible route would be via use of `importlib`, of which Access object library does a fair number, but still requires that the VBA project add references for the imported type libraries to get early-binding for those imported types. Another route is to embed multiple type libraries into the same output file. The downside is that you'd have to use syntax like `...\foo\bar\myTBProject.dll\3` -- what is `3`? How is the `3` different from the `2`? Or the default one for that matter?

For that reason, there's a lot of motivation toward dumping all the types into one giant type library as to avoid having to reference each dependencies. The ideal solution should help avoid that from causing problems especially when we raise the level of abstraction in a tB project.

If we choose to make namespace a logical organization units, for use within tB, independent from the type libraries that is generated, then we have considerably more freedom in defining how it should work. The downside is that the author then has to maintain 2 parallel organizations; the type library representation and the namespace organization. 

I can see nested class buckets working as long we keep to the convention that the bucket is not creatable & cannot be referenced. However, that requires it to be a module, rather than a class and as far as I know, a module cannot contain another module, so that limits the nesting to only one level. Furthermore, it would feel weird to see a "class" in the object browser that is nothing like an actual class. 

If we cannot avoid the limitations in the scoping for the type libraries as defined in the MS-OAUT & MS-VBAL, my vote would be to make the namespace a tB-only construct for aid in logical organization. We should easily control the type library's shape via use of attributes. The extra responsibility of maintaining 2 different scoping hierarchies is an acceptable trade-off for me. That is because the feature of being able to control what is in the global namespace and how to disambiguate the types is more important to me especially for private implementation details.

**Comment by WindowStations** [comment]
In a larger namespace setup, you can't have multiple references without adding a fair amount of difficulty.  Many classes rely on other classes already which makes it difficult to model/manage both for the developer and end-developer.  But, remember that you MUST also integrate these class hierarchies with the UI Control set you're designing for Windows.  These things work together and will be designed together when they overlap.  Otherwise you need extra code to handle missing references etc.

Also, note the property I posted above can be declared as private in individual modules.  Then, other modules won't see IO.

Would TB be able to offer the same range of quantity/quality content available through com-visible NET namespaces?
Would TB's namespaces be built by scratch, or are they exposed in the model?

**Comment by WindowStations** [comment]
Global naming collisions don't really happen if you explicitly instance the classes.
  Dim p1 As New VBCtl.Process' Namespace class from comvisible NET dll
  Dim p2 As New Project1.Process 'Project class .cls
  Dim p3 As New RefencencedLib.Process 'referenced class dll
These are different instances of classes with the same name.  How is this a limitation again?

**Comment by bclothier** [comment]
Just to take the discussion from #634  and put it where it belongs. I think that is relevant here because this touches on shadowing which overlaps the problem domain that the namespaces solves.

There seem to be a consensus that we should have the ability to:

1) Unreference the default `VBA` (and other related libraries) altogether
2) Put some other library as higher precedence over the `VBA`. 

Allowing either would then allow anybody to effectively create their own standard library and even shadow the common functions with their custom implementation. 

In VBx, it is not possible to write a project that shadows the existing functions and use it in other project. One consequence of that is that it encourages copy pasta, especially in VBA which then get exacerbated when the copied code is then altered, so we now have 10 different versions of shadowing functions. We want to make it easy to publish a library and use it in other projects.  The current VBx way adds the reference to the global namespace and use precedence to resolve ambiguity. In a complex project, that can result in a very polluted intellisense listing all unrelated functions that are not relevant to what you are coding at the moment. `appobject`s are other big headaches; adding references to type libraries that contains `appobject`s that has similar methods can cause problem especially if the access is unqualified. This is a real problem that need a solution. 

We need to consider whether shadowing is a good idea, especially for common standard functions. I have less problem with providing a custom implementation of an interface than I do with a global function because the latter violates the principle of least astonishment. It would make for a frustrating experience if I was working in someone else's project and it was not clear to me that the `StrConv` was shadowed by a custom implementation, especially because VBx does not have `using` or `import` statement. Moreover, it really does not address the separate problem of the global namespace pollution.

I'm intended to think that hierarchical namespaces is still a good idea with or without the ability to shadow the standard libraries. Just to look at what other languages has done, they do similar things. Python has [Import](https://docs.python.org/3/reference/import.html). So does Rust using the [`mod` syntax](http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch07-01-mod-and-the-filesystem.html). Looking at the rust's example, I could see doing something like this:

```
Public Module MyBigGnarlyStringLibrary
  Public Module Unicode
     ...
  End Module
  
  Public Module ANSI
    ...
  End Module
End Module
```

The problem is that it would not be COM-compatible because modules cannot be nested the way we can with interfaces. It might be acceptable for use within tB coding, but if we allow that, then we need to then deal with how to export them in the published type library.
